chalk-solve/src/clauses/builtin_traits/fn_.rs

```diff
@@ -0,0 +1,93 @@
+use crate::clauses::ClauseBuilder;
+use crate::infer::instantiate::IntoBindersAndValue;
+use crate::rust_ir::WellKnownTrait;
+use crate::{Interner, RustIrDatabase, TraitRef};
+use chalk_ir::{
+    AliasTy, ApplicationTy, Binders, Normalize, ProjectionTy, Substitution, TraitId, Ty, TyData,
+    TypeName, VariableKinds,
+};
+
+/// Handles clauses for FnOnce/FnMut/Fn.
+/// If `assoc_output` is `true`, we push a clause of the form
+/// `Normalize(<fn(A) -> B as FnOnce<(A,)>>::Output -> B) :- Implemented(fn(A) -> B as FnOnce<(A,)>`
+///
+/// If `assoc_output` is `false`, we push a clause of the form
+/// `Implemented(fn(A) -> B as FnOnce<(A,)>)`
+pub fn add_fn_trait_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_id: TraitId<I>,
+    ty: &TyData<I>,
+    assoc_output: bool,
+) {
+    match ty {
+        TyData::Function(fn_val) => {
+            let interner = db.interner();
+            let (binders, orig_sub) = fn_val.into_binders_and_value(interner);
+            // Take all of the arguments except for the last one, which
+            // represents the return type
+            let arg_sub = Substitution::from(
+                interner,
+                orig_sub.iter(interner).take(orig_sub.len(interner) - 1),
+            );
+            let fn_output_ty = orig_sub
+                .at(interner, orig_sub.len(interner) - 1)
+                .assert_ty_ref(interner);
+
+            // We are constructing a reference to `FnOnce<Args>`, where
+            // `Args` is a tuple of the function's argument types
+            let tupled = Ty::new(
+                interner,
+                TyData::Apply(ApplicationTy {
+                    name: TypeName::Tuple(arg_sub.len(interner)),
+                    substitution: arg_sub.clone(),
+                }),
+            );
+
+            let self_ty = Ty::new(interner, ty);
+
+            let tupled_sub = Substitution::from(interner, vec![self_ty, tupled]);
+            // Given a function type `fn(A1, A2, ..., AN)`, construct a `TraitRef`
+            // of the form `fn(A1, A2, ..., AN): FnOnce<(A1, A2, ..., AN)>`
+            let new_trait_ref = TraitRef {
+                trait_id,
+                substitution: tupled_sub.clone(),
+            };
+
+            // Functions types come with a binder, which we push so
+            // that the `TraitRef` properly references any bound lifetimes
+            // (e.g. `for<'a> fn(&'a u8): FnOnce<(&'b u8)>`)
+            let bound_ref = Binders::new(VariableKinds::from(interner, binders), new_trait_ref);
+            builder.push_binders(&bound_ref, |this, inner_trait| {
```

Comment:
This seems wrong to do this *here*. Instead, I would move this to the top of the function.

So the top of the function would look something like
```
let interner = db.interner();
let (binders, orig_sub) = fn_val.into_binders_and_value(interner);
let bound_ref = Binders::new(VariableKinds::from(interner, binders), orig_sub);
builder.push_binders(&bound_ref, |builder, orig_sub| {
```

Then, everything else would be mostly the same. And `inner_trait` here would just be the `TraitRef` you create right above (`new_trait_ref`).

Reply:
Side note: I definitely prefer the `this` -> `builder`, since it means we can't accidentally reference the outer builder from within.